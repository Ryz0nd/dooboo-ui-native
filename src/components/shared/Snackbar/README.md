# Snackbar

[![Npm Version](http://img.shields.io/npm/v/@dooboo-ui/native-snackbar.svg?style=flat-square)](https://npmjs.org/package/@dooboo-ui/native-snackbar)
[![Downloads](http://img.shields.io/npm/dm/@dooboo-ui/native-snackbar.svg?style=flat-square)](https://npmjs.org/package/@dooboo-ui/native-snackbar)

> Simple snackbar for react-native.

## Installation

```sh
yarn add @dooboo-ui/native
```

or

```sh
yarn add @dooboo-ui/native-snackbar
```

## Usage

# Props

```ts
export enum Timer {
  SHORT = 1500,
  LONG = 3000,
}

export interface SnackbarProps {
  text: string;
  testID?: string;
  show: boolean;
  setShow: (show: boolean) => void;
  timer?: Timer;
}
```

|         | necessary | types                   | default   | info         |
| ------- | :-------: | ----------------------- | --------- | ------------ |
| text    |     ✓     | string                  | undefined |              |
| testID  |           | string                  | undefined |              |
| timer   |           | number                  | 1500      | closing time |
| show    |     ✓     | boolean                 | undefined |              |
| setSHow |     ✓     | (show: boolean) => void | undefined |              |

# Getting started

- Import

  ```tsx
  import { Snackbar } from '@dooboo-ui/native';
  // or
  import Snackbar from '@dooboo-ui/native-snackbar';
  ```

- Usage
  ```tsx
  const SnackbarWithState = () => {
    const [show, setShow] = useState<boolean>(false);
    const [timer, setTimer] = useState<Timer>(Timer.SHORT);

    return (
      <Container>
        <Button
          style={{
            marginVertical: 40,
          }}
          text="OPEN SNACKBAR (SHORT TIMER)"
          onClick={(): void => {
            setShow(true);
            setTimer(Timer.SHORT);
          }}
        />
        <Button
          text="OPEN SNACKBAR (LONG TIMER)"
          style={{
            marginVertical: 40,
          }}
          onClick={(): void => {
            setShow(true);
            setTimer(Timer.LONG);
          }}
        />
        <Snackbar
          text={text('Snackbar Text', 'Simple Snackbar is opened')}
          show={show}
          setShow={setShow}
          timer={timer}
        />
      </Container>
    );
  };
  ```
