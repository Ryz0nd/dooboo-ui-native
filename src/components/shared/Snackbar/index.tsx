import * as React from 'react';

import { Text, TextStyle, ViewStyle, Dimensions } from 'react-native';
import styled from 'styled-components/native';

const { width } = Dimensions.get('screen');
const maxWidth = width - 32;

const Container = styled.View`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  min-width: 150px;
  max-width: ${maxWidth};
  text-align: left;
  position: absolute;
  font-size: 16;
  padding: 16px;
  bottom: 50px;
  background-color: #303235;
  border-radius: 10;
`;

const ActionContainer = styled.View`
  display: flex;
  align-items: center;
  margin-left: auto;
  padding-left: 16px;
`;

export interface SnackbarProps {
  text: string;
  testID?: string;
  show: boolean;
  setShow: (show: boolean) => void;
  timer?: Timer;
  action?: React.ReactNode;
  containerStyle?: ViewStyle;
  messageStyle?: TextStyle;
}

export enum Timer {
  SHORT = 1500,
  LONG = 3000,
}

const Snackbar: React.FC<SnackbarProps> = (props) => {
  const { testID, show, setShow, timer = Timer.SHORT, action } = props;
  React.useEffect(() => {
    let timeout;
    if (show === true) {
      timeout = setTimeout(() => {
        setShow(false);
      }, timer);
    }
    return (): void => clearTimeout(timeout);
  }, [show]);
  return (
    <>
      {show && (
        <Container testID={testID}
          style={props.containerStyle}>
          <Text
            style={props.messageStyle}
          >
            {props.text}
          </Text>
          {action && <ActionContainer>{action}</ActionContainer>}
        </Container>
      )}
    </>
  );
};

export default Snackbar;
