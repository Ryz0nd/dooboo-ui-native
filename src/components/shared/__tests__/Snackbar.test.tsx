import 'react-native';

import * as React from 'react';

import Snackbar, { SnackbarProps } from '../Snackbar';

import { render } from '@testing-library/react-native';
import renderer from 'react-test-renderer';

// Note: test renderer must be required after react-native.

const props: SnackbarProps = {
  testID: 'snackbar',
  text: 'SnackBar!',
  show: true,
  setShow: jest.fn(),
};
let component: React.ReactElement;

jest.useFakeTimers();

describe('[Snackbar]', () => {
  beforeEach(() => {
    component = <Snackbar {...props} />;
  });

  it('renders without crashing', () => {
    const rendered = renderer.create(component);
    expect(rendered.toJSON()).toMatchSnapshot();
    expect(rendered.toJSON()).toBeTruthy();

    rendered.update(<Snackbar {...{ ...props, show: false }} />);
    expect(rendered.toJSON()).toBeFalsy();
  });

  it('should dismiss after timeout', async () => {
    render(component);
    jest.runAllTimers();
    expect(props.setShow).toBeCalled();
  });
});
