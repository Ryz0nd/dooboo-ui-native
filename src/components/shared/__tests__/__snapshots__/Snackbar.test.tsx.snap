// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Snackbar] renders without crashing 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#303235",
        "borderRadius": 10,
        "bottom": 50,
        "display": "flex",
        "flexDirection": "row",
        "flexWrap": "wrap",
        "fontSize": 16,
        "maxWidth": 718,
        "minWidth": 150,
        "paddingBottom": 10,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 10,
        "position": "absolute",
        "textAlign": "left",
      },
    ]
  }
  testID="snackbar"
>
  <Text>
    SnackBar!
  </Text>
</View>
`;
